subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/AbstractCacheCleanup.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public void execute(final PersistentCache persistentCache) {
+        buildOperationExecutor.run(new RunnableBuildOperation() {
+            @Override
+            public void run(BuildOperationContext context) {
+                cleanup(persistentCache);
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Clean up " + persistentCache);
+            }
+        });
+
+    }
+
+    private void cleanup(final PersistentCache persistentCache) {
+        final File[] filesEligibleForCleanup = buildOperationExecutor.call(new CallableBuildOperation<File[]>() {
+            @Override
+            public File[] call(BuildOperationContext context) {
+                return findEligibleFiles(persistentCache.getBaseDir());
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Scan " + persistentCache.getBaseDir());
+            }
+        });
+
+        if (filesEligibleForCleanup.length > 0) {
+            final List<File> filesForDeletion = buildOperationExecutor.call(new CallableBuildOperation<List<File>>() {
+                @Override
+                public List<File> call(BuildOperationContext context) {
+                    return findFilesToDelete(persistentCache, filesEligibleForCleanup);
+                }
+
+                @Override
+                public BuildOperationDescriptor.Builder description() {
+                    return BuildOperationDescriptor.displayName("Choose files to delete from " + persistentCache);
+                }
+            });
+
+            if (!filesForDeletion.isEmpty()) {
+                buildOperationExecutor.run(new RunnableBuildOperation() {
+                    @Override
+                    public void run(BuildOperationContext context) {
+                        cleanupFiles(persistentCache, filesForDeletion);
+                    }
+
+                    @Override
+                    public BuildOperationDescriptor.Builder description() {
+                        return BuildOperationDescriptor.displayName("Delete files for " + persistentCache);
+                    }
+                });
+            }
+        }
+    }
+
+    protected abstract List<File> findFilesToDelete(final PersistentCache persistentCache, File[] filesEligibleForCleanup);
+
+    File[] findEligibleFiles(File cacheDir) {
+        // TODO: This doesn't descend subdirectories.
+        return cacheDir.listFiles(new FilenameFilter() {
+            @Override
+            public boolean accept(File dir, String name) {
+                return canBeDeleted(name);
+            }
+        });
+    }
+
+    protected boolean canBeDeleted(String name) {
+        return !(name.endsWith(".properties") || name.endsWith(".lock"));
```

Comment:
it would be better to ask the `PersistentCache` which files it owns instead of assuming a particular implementation.

Reply:
Good idea